src/python/pants/backend/visibility/rule_types_test.py

```diff
@@ -212,67 +220,113 @@ def dependents_rules() -> BuildFileVisibilityRules:
-        (parse_ruleset(("*", ("src/ok/*", "?src/dubious/*", "!src/blocked/*")), "tgt/BUILD"),),
+        (
+            parse_ruleset(("requirement", "*"), "BUILD"),
+            parse_ruleset(("*", ("src/ok/*", "?src/dubious/*", "!src/blocked/*")), "tgt/BUILD"),
+        ),
     )
 
 
 @pytest.mark.parametrize(
-    "source_path, target_path, expected_action, expected_rule",
+    "target_type, source_path, target_path, expected_action, expected_rule",
     [
-        ("src/ok/a", "tgt/ok/b", "allow", "src/BUILD[tgt/ok/*] -> tgt/BUILD[src/ok/*]"),
-        ("src/ok/a", "tgt/dubious/b", "warn", "src/BUILD[?tgt/dubious/*] -> tgt/BUILD[src/ok/*]"),
-        ("src/ok/a", "tgt/blocked/b", "deny", "src/BUILD[!tgt/blocked/*] -> tgt/BUILD[src/ok/*]"),
-        ("src/dubious/a", "tgt/ok/b", "warn", "src/BUILD[tgt/ok/*] -> tgt/BUILD[?src/dubious/*]"),
+        ("test", "src/ok/a", "tgt/ok/b", "allow", "src/BUILD[tgt/ok/*] -> tgt/BUILD[src/ok/*]"),
+        (
+            "test",
+            "src/ok/a",
+            "tgt/dubious/b",
+            "warn",
+            "src/BUILD[?tgt/dubious/*] -> tgt/BUILD[src/ok/*]",
+        ),
+        (
+            "test",
+            "src/ok/a",
+            "tgt/blocked/b",
+            "deny",
+            "src/BUILD[!tgt/blocked/*] -> tgt/BUILD[src/ok/*]",
+        ),
         (
+            "test",
+            "src/dubious/a",
+            "tgt/ok/b",
+            "warn",
+            "src/BUILD[tgt/ok/*] -> tgt/BUILD[?src/dubious/*]",
+        ),
+        (
+            "test",
             "src/dubious/a",
             "tgt/dubious/b",
             "warn",
             "src/BUILD[?tgt/dubious/*] -> tgt/BUILD[?src/dubious/*]",
         ),
         (
+            "test",
             "src/dubious/a",
             "tgt/blocked/b",
             "deny",
             "src/BUILD[!tgt/blocked/*] -> tgt/BUILD[?src/dubious/*]",
         ),
-        ("src/blocked/a", "tgt/ok/b", "deny", "src/BUILD[tgt/ok/*] -> tgt/BUILD[!src/blocked/*]"),
         (
+            "test",
+            "src/blocked/a",
+            "tgt/ok/b",
+            "deny",
+            "src/BUILD[tgt/ok/*] -> tgt/BUILD[!src/blocked/*]",
+        ),
+        (
+            "test",
             "src/blocked/a",
             "tgt/dubious/b",
             "deny",
             "src/BUILD[?tgt/dubious/*] -> tgt/BUILD[!src/blocked/*]",
         ),
         (
+            "test",
             "src/blocked/a",
             "tgt/blocked/b",
             "deny",
             "src/BUILD[!tgt/blocked/*] -> tgt/BUILD[!src/blocked/*]",
         ),
+        (
+            "requirement",
+            "src/proj/code.ext",
+            "3rdparty/req#lib",
+            "allow",
+            "BUILD[//3rdparty/**] -> BUILD[*]",
+        ),
+        (
+            "requirement",
+            "src/proj/code.ext",
+            "3rdparty/req#restricted",
+            "deny",
+            "BUILD[!//3rdparty/req#restrict*] -> BUILD[*]",
```

Comment:
Once you get used to reading these, they're actually really helpful.
> "BUILD[!//3rdparty/req#restrict*] -> BUILD[*]",

The format is:
```
   "BUILD[      !//3rdparty/req#restrict*         ] ->  BUILD[   *  ]",
origin^  matching dependency rule ^ for origin, dependency^      ^matching dependents rule for the dependency
```